cmd/egress-cni-plugin/main.go

```diff
@@ -0,0 +1,183 @@
+
+package main
+
+import (
+	"fmt"
+	"runtime"
+	"strconv"
+
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/iptableswrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/vethwrapper"
+
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/hostipamwrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/netlinkwrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/nswrapper"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/procsyswrapper"
+
+	"github.com/coreos/go-iptables/iptables"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	cniversion "github.com/containernetworking/cni/pkg/version"
+	"github.com/containernetworking/plugins/pkg/utils"
+)
+
+var version string
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func main() {
+	skel.PluginMain(cmdAdd, nil, cmdDel, cniversion.All, fmt.Sprintf("egress CNI plugin %s", version))
+}
+
+func cmdAdd(args *skel.CmdArgs) (err error) {
+	iptCreator := func(protocol iptables.Protocol) (iptableswrapper.IPTablesIface, error) {
+		return iptableswrapper.NewIPTables(protocol)
+	}
+	c := &EgressContext{
+		Procsys:    procsyswrapper.NewProcSys(),
+		Ns:         nswrapper.NewNS(),
+		NsPath:     args.Netns,
+		ArgsIfName: args.IfName,
+		Ipam:       hostipamwrapper.NewIpam(),
+		Link:       netlinkwrapper.NewNetLink(),
+		Veth:       vethwrapper.NewSetupVeth(),
+		IptCreator: iptCreator,
+	}
+
+	return _cmdAdd(args, c)
+}
+
+func _cmdAdd(args *skel.CmdArgs, c *EgressContext) (err error) {
+	c.NetConf, c.Log, err = LoadConf(args.StdinData)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+	if c.NetConf.PrevResult == nil {
+		c.Log.Debugf("must be called as a chained plugin")
+		return fmt.Errorf("must be called as a chained plugin")
+	}
+
+	c.Result, err = current.GetResult(c.NetConf.PrevResult)
+	if err != nil {
+		c.Log.Errorf("failed to get PrevResult: %v", err)
+		return err
+	}
+	// Convert MTU from string to int
+	c.Mtu, err = strconv.Atoi(c.NetConf.MTU)
+	if err != nil {
+		c.Log.Errorf("failed to parse MTU: %s, err: %v", c.NetConf.MTU, err)
+		return err
+	}
+
+	c.Log.Debugf("Received an ADD request for: conf=%v; Plugin enabled=%s", c.NetConf, c.NetConf.Enabled)
+	// We will not be vending out this as a separate plugin by itself, and it is only intended to be used as a
+	// chained plugin to VPC CNI. We only need this plugin to kick in if egress is enabled in VPC CNI. So, the
+	// value of an env variable in VPC CNI determines whether this plugin should be enabled and this is an attempt to
+	// pass through the variable configured in VPC CNI.
+	if c.NetConf.Enabled != "true" {
+		return types.PrintResult(c.Result, c.NetConf.CNIVersion)
+	}
+
+	if c.IpTablesIface == nil {
```

Comment:
nit: instead of having `IptCreator` as a field, we can just set `c.IpTablesIface` here. I don't think the lambda function simplifies anything, as we don't need to pass anything to the constructor

Reply:
I think the key here is for unit testing. Unit testing requires that this field to be injected into our func, and if we assign value here directly here, it will overwrite the value passed by unit testing. 

By using lambda and if c.IpTablesIface == nil statement, we achieved:
1. iptables will not be instanced until last step - for performance reason.
2. with `if` statement, unit testing will directly pass value to c.IpTablesIface and its value will not be overwriten. Unit testing has to setup bunch of EXPECT statement on this c.IpTablesIface before running the code, so we have to  have this `if` statement to prevent that.
3. So, for unit testing, the c.IpTablesIface is initiated before code starts, and in real situation, this field is initiated later on.

Reply:
That's a great point, I did not think about ease of mocking. Resolving this